docs/design/secrets.md

```diff
@@ -0,0 +1,194 @@
+SaaS API tokens, etc.  Users should be able to consume configuration artifacts in their containers and
+be able to control the path on the container's filesystems where the artifact will be presented.
+
+### Use-Case: Metadata about services
+
+Most pieces of information about how to use a service are secrets.  For example, a service that
+provides a MySQL database needs to provide the username, password, and database name to consumers
+so that they can authenticate and use the correct database. Containers in pods consuming the MySQL
+service would also consume the secrets associated with the MySQL service.
+
+## Deferral: Consuming secrets as environment variables
+
+Some images will expect to receive configuration items as environment variables instead of files.
+We should consider what the best way to allow this is; there are a few different options:
+
+1.  Force the user to adapt files into environment variables.  Users can store secrets that need to
+    be presented as environment variables in a format that is easy to consume from a shell:
+
+        $ cat /etc/secrets/my-secret.txt
+        export MY_SECRET_ENV=MY_SECRET_VALUE
+
+    The user could `source` the file at `/etc/secrets/my-secret` prior to executing the command for
+    the image either inline in the command or in an init script, 
+
+2.  Give secrets an attribute that allows users to express the intent that the platform should
+    generate the above syntax in the file used to present a secret.  The user could consume these
+    files in the same manner as the above option.
+
+3.  Give secrets attributes that allow the user to express that the secret should be presented to
+    the container as an environment variable.  The container's environment would contain the
+    desired values and the software in the container could use them without accomodation the
+    command or setup script.
+
+For our initial work, we will treat all secrets as files to narrow the problem space.  There will
+be a future proposal that handles exposing secrets as environment variables.
+
+## Flow analysis of secret data
+
+There are two fundamentally different use-cases for access to secrets:
+
+1.  CRUD operations on secrets by their owners
+2.  Read-only access to the secrets needed for a particular node by the kubelet
+
+### Use-Case: CRUD operations by owners
+
+In use cases for CRUD operations, the user experience for secrets should be no different than for
+other API resources.
+
+TODO: document possible relation with Service Accounts
+
+### Use-Case: Kubelet read of secrets for node
+
+The use-case where the kubelet reads secrets has several additional requirements:
+
+1.  Kubelets should only receive secret data which is required by pods scheduled onto the kubelet's
+    node
+2.  Kubelets should have read-only access to secret data
+3.  Secret data should not be transmitted over the wire insecurely
```

Comment:
Should secret data be unencrypted at rest on the node?   Multiple choice:
  1.  No.  (Needs discussion around threats and what types of secrets are appropriate to store this way)
  2.  Yes.  (how to do this?  make an encrypted loop device?)
  3. Not applicable.  The data never is at rest.  Always put in a tmpfs.  (Is size limitation on secrets okay?)